### PR TITLE
🔴🔴🔴 [`Attention`] Refactor Attention Interface for Bart-based Models

### PR DESCRIPTION
**Pull request by vasqu**
This PR is gonna tackle two things in general:
- [x] Flex Attention for all base attention types (encoder, decoder, encoder-decoder cross)
- [x] New Attention Interface for a bunch of models (mostly based on Bart's implementation)
- [x] As a bonus, some models have already been refactored into modular if seen fit, e.g. PL Bart.

Affected models (will be updated when I have enough time) - probably not 100% accurate
- [x] Bart
- [x] Biogpt
- [x] Mbart
- [x] Bigbird Pegasus
- [x] Blenderbot
- [x] Blenderbot Small
- [x] Wav2vec
- [x] Data2vec audio
- [x] Informer
- [x] Timeseries Transformer
- [x] Hubert
- [x] M2M 100
- [x] Marian
- [x] Musicgen
- [x] Music Melody
- [x] Nllb MoE
- [x] Patchts Mixer
- [x] Patchtst
- [x] Pegasus
- [x] Pegasus X
- [x] PL Bart
- [x] Sew 
- [x] Speech to Text 
- [x] Uni Speech 
- [x] Uni Speech SAT 

Possibly doable in this PR:
- [ ] Rename flex attn mask creation since the focus is currently only on decoder-only ones, the naming is not really fitting 

Worth a discussion (?):
- Time Series Transformer and Speech to Text theoretically have the new attentions but the tests are unsuitable (as they generate inputs with ids). Imo, a rewrite/adjustment is too much effort. I disabled it for now but we could also enable them and give the users a warning (untested).
- Pegasus X has more flaky logits on sdpa - should it still be enabled?
- Nllb MoE has issues with flash attention as masks are prepared differently and used by the MoE leading to several issues + sdpa has also more flaky logits --> I think it's not worth it based on the size of the model + the usage.

Worth a discussion TL;DR:
- Timer Series Transformer + Speech to Text do not work with the current testing framework (need special attention to what they get as input)
- Pegasus X has more flaky logits on sdpa - should it still be enabled?
- Nllb MoE is too complicated to enable fastly for Flash Attention (attention mask <-> MoE interaction) + same as pegasus x on sdpa logits.

Future PRs will address:
- [ ] Other models such as Bert-based models.
- [ ] Other models such as Whisper-based models.
- [ ] More modular.
- [ ] Proper kwargs passing --> multiple attn types should get different kwargs?
- [ ] In combination with above point ^ optionally tp plans etc for attn backend support?
- [ ] Fa might be able to work without position ids but fa_kwargs only - this is a current limitation in the fa modeling utils...
- [ ] Mask refactor as iirc the prepare_for... are not really neat. However, the way it is written, refactoring should be easy ;)
- [ ] Flex Attn tests and fixes (compile issues)
- [ ] Better datasets versioning? Pipeline issues esp with audio

### CODE REVIEW

File: src/transformers/models/autoformer/modeling_autoformer.py
Comment (ArthurZucker) [inline src/transformers/models/autoformer/modeling_autoformer.py:1]: attention api is being refactored so this will go away anyways with #37866

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -198,316 +232,49 @@ def forward(
...
+        attention_type = self.config._attn_implementation
+        if self.config._attn_implementation != "eager":
+            if (output_attentions or layer_head_mask is not None) and self.config._attn_implementation in [
+                "sdpa",
+                "flash_attention_2",
+                "flex_attention",
+            ]:
+                logger.warning_once(
+                    f"Falling back to eager attention because `{attention_type}` does not support"
+                    f" `output_attentions=True` or `head_mask`."
```
Comment (ArthurZucker) [inline src/transformers/models/bart/modeling_bart.py:245]: I think we can put the warning in the attention interface functions themselves!

File: src/transformers/models/bart/modeling_bart.py [hunk 2]
```diff
@@ -885,18 +651,12 @@ def forward(
...
-                attention_mask = attention_mask if 0 in attention_mask else None
-            elif self._use_sdpa and head_mask is None and not output_attentions:
-                # output_attentions=True & head_mask can not be supported when using SDPA, fall back to
-                # the manual implementation that requires a 4D causal mask in all cases.
-                # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
-                attention_mask = _prepare_4d_attention_mask_for_sdpa(attention_mask, inputs_embeds.dtype)
-            else:
-                # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
-                attention_mask = _prepare_4d_attention_mask(attention_mask, inputs_embeds.dtype)
+        _unsupported_features = output_attentions is True or head_mask is not None
```
Comment (ArthurZucker) [inline src/transformers/models/bart/modeling_bart.py:654]: not a fan of this as it won't be clear why they are not supported for users! mask is getting refactored anyways!

File: src/transformers/models/bart/modeling_bart.py [hunk 3]
```diff
@@ -1103,42 +888,21 @@ def forward(
...
+            past_key_values_length,
+            _unsupported_features,
+        )
+        encoder_attention_mask = self._update_cross_attn_mask(
+            encoder_hidden_states,
+            encoder_attention_mask,
+            input_shape,
+            inputs_embeds,
+            _unsupported_features,
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/bart/modeling_bart.py:905]: not bad!

File: src/transformers/models/biogpt/modular_biogpt.py
Comment (ArthurZucker) [inline src/transformers/models/biogpt/modular_biogpt.py:1]: super nice!

File: src/transformers/models/blenderbot/modeling_blenderbot.py [hunk 4]
```diff
@@ -186,72 +231,41 @@ def forward(
...
+                "flex_attention",
+            ]:
+                logger.warning_once(
+                    f"Falling back to eager attention because `{attention_type}` does not support"
+                    f" `output_attentions=True` or `head_mask`."
                 )
-            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + attention_mask
-            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
-
-        attn_weights = nn.functional.softmax(attn_weights, dim=-1)
```
Comment (ArthurZucker) [inline src/transformers/models/blenderbot/modeling_blenderbot.py:211]: I think we need to be careful about the dtype, here the softmax is not in float32

File: src/transformers/models/plbart/modular_plbart.py [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+    BartDecoder,
+    BartEncoder,
+    BartForCausalLM,
+    BartScaledWordEmbedding,
+)
+from .configuration_plbart import PLBartConfig
+
+
+# Copied from transformers.models.mbart.modeling_mbart.shift_tokens_right
+def shift_tokens_right(input_ids: torch.Tensor, pad_token_id: int):
```
Comment (ArthurZucker) [inline src/transformers/models/plbart/modular_plbart.py:45]: if copied from, you can import it!

File: src/transformers/models/plbart/modular_plbart.py [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+    pass
+
+
+@auto_docstring(
+    custom_intro="""
+    PLBart model with a sequence classification/head on top (a linear layer on top of the pooled output) e.g. for code
+    classification.
+    """
+)
+class PLBartForSequenceClassification(PLBartPreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/plbart/modular_plbart.py:435]: forward of this one seems the same as big bird no?

File: src/transformers/models/sew/modular_sew.py [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+    Wav2Vec2NoLayerNormConvLayer,
+    Wav2Vec2SamePadLayer,
+)
+from .configuration_sew import SEWConfig
+
+
+_HIDDEN_STATES_START_POSITION = 1
+
+
+# Copied from transformers.models.wav2vec2.modeling_wav2vec2._compute_mask_indices
```
Comment (ArthurZucker) [inline src/transformers/models/sew/modular_sew.py:50]: if it is copied from it can be imported

File: src/transformers/models/time_series_transformer/modeling_time_series_transformer.py [hunk 8]
```diff
@@ -358,69 +404,41 @@ def forward(
...
+                "flex_attention",
+            ]:
+                logger.warning_once(
+                    f"Falling back to eager attention because `{attention_type}` does not support"
+                    f" `output_attentions=True` or `head_mask`."
                 )
-            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + attention_mask
-            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
-
-        attn_weights = nn.functional.softmax(attn_weights, dim=-1)
```
Comment (ArthurZucker) [inline src/transformers/models/time_series_transformer/modeling_time_series_transformer.py:383]: careful again as the default is not float32

File: tests/models/bart/test_modeling_bart.py [hunk 9]
```diff
@@ -1534,3 +1534,7 @@ def test_decoder_model_attn_mask_past(self):
     @unittest.skip(reason="Decoder cannot keep gradients")
     def test_retain_grad_hidden_states_attentions(self):
         return
+
+    @unittest.skip(reason="Decoder cannot keep gradients")
```
Comment (ArthurZucker) [inline tests/models/bart/test_modeling_bart.py:1538]: a nice todo!

File: tests/models/m2m_100/test_modeling_m2m_100.py [hunk 10]
```diff
@@ -439,7 +439,7 @@ def test_flash_attn_2_seq_to_seq_generation(self):
         Overwriting the common test as the test is flaky on tiny models
         """
         model = M2M100ForConditionalGeneration.from_pretrained(
-            "facebook/m2m100_418M", attn_implementation="flash_attention_2"
+            "facebook/m2m100_418M", attn_implementation="flash_attention_2", torch_dtype=torch.float16
```
Comment (ArthurZucker) [inline tests/models/m2m_100/test_modeling_m2m_100.py:442]: we should not have to change this!@

**Review by ArthurZucker — COMMENTED** [review]
Very nice! be careful about softmax dtype, and let's put the warning in the attention interface to reduce the visibility of incompatibility!

File: src/transformers/models/blenderbot/modeling_blenderbot.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/blenderbot/modeling_blenderbot.py:211]: Good catch, changed the behaviour back to the old one!

File: src/transformers/models/plbart/modular_plbart.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/models/plbart/modular_plbart.py]: Yup, changed it to inheret from it :) even smaller modular now

File: tests/models/m2m_100/test_modeling_m2m_100.py (hunk 10 above)
Comment (vasqu) [reply tests/models/m2m_100/test_modeling_m2m_100.py:442]: Also fails on main without this change. Seems like this test has been broken for a while - flash attention would get `float32` as input which leads to it failing. Not sure if some old loading logic adjusted the dtype automatically to half.

File: src/transformers/models/bart/modeling_bart.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py:245]: I agree, we should move them tbh. But, it follows the logic in llama (which supports less of the features like `head_mask` https://github.com/huggingface/transformers/blob/d69945e5fced637e77cf7af5e4955cb897bc298c/src/transformers/models/llama/modeling_llama.py#L256-L263

Since this will affect all models using some sort of the interface, I think it's smarter to keep it for now and add a wrapper for this in a different PR? Wdyt?

File: src/transformers/models/bart/modeling_bart.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py:654]: Should I add a comment here? I feel like users will get the appropriate warning within the attention module then (output_attentions / head_mask). 

But yes, let's refactor it in a different PR when masking etc is appropriately handled.

**Comment by vasqu** [comment]
Addressed the dtype issue and the modular ones. 

For the warnings, I think it makes more sense to create another PR that wraps the current interface way since it affects more models than just here.

**Comment by vasqu** [comment]
Discovered some issues in the flex attn tests due to incompatible kwargs - investigating :eyes: mostly audio models with different kwargs

**Comment by vasqu** [comment]
Merge conflict with #35314 is very huge while changing previous dependencies - keeping track of the models again while working locally:
- [x] Bart
- [ ] Biogpt
- [ ] Mbart
- [x] Bigbird Pegasus
- [ ] Blenderbot
- [ ] Blenderbot Small
- [ ] Wav2vec
- [ ] Data2vec audio
- [ ] Informer
- [ ] Timeseries Transformer
- [ ] Hubert
- [ ] M2M 100
- [ ] Marian
- [ ] Musicgen
- [ ] Music Melody
- [ ] Nllb MoE
- [ ] Patchts Mixer
- [ ] Patchtst
- [ ] Pegasus
- [ ] Pegasus X
- [ ] PL Bart
- [ ] Sew 
- [ ] Speech to Text 
- [ ] Uni Speech 
- [ ] Uni Speech SAT

File: src/transformers/models/bart/modeling_bart.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/bart/modeling_bart.py:245]: Yes and no! It's "alright" if its hidden outside the modeling, I'd prefer we unclutterthe modeling!

**Comment by ArthurZucker** [comment]
Okay, good luck it looks tough!

File: src/transformers/models/bart/modeling_bart.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py]: Moved all that to the interface in modeling_utils:
https://github.com/huggingface/transformers/blob/ece5b09003cb247ede738536dfa795d4eab08531/src/transformers/modeling_utils.py#L6142-L6194

Makes the modeling files definitely less bloated. Also addresses #38123 when future models are refactored with this usage; focussed on this set of models first.

File: src/transformers/models/bart/modeling_bart.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py:866]: Added comments

File: src/transformers/models/bart/modeling_bart.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py]: cc @molbap for #38123 see the defaulting for eager

**Comment by vasqu** [comment]
PR is ready again, will look whether some flags have been missed but it's good otherwise.

**Review by ArthurZucker — COMMENTED** [review]
🔴 for the abstraction of the attention interface! 
Let's keep the core logic explicit, and just put each warning in each integration/sdpa|flash|flex.py file!

File: src/transformers/models/bart/modeling_bart.py [hunk 11]
```diff
@@ -244,19 +252,18 @@ def forward(
...
-            attention_type=self.config._attn_implementation,
-            eager_attention=eager_attn_forward,
-            module=self,
-            query=query_states,
-            key=key_states,
-            value=value_states,
-            attention_mask=attention_mask,
-            training=self.training,
-            dropout=self.dropout,
+        attention_interface: Callable = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
Comment (ArthurZucker) [inline src/transformers/models/bart/modeling_bart.py]:
```suggestion
        attention_interface = eager_attn_forward
        if self.config._attn_implementation != "eager":
          attention_interface: Callable = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
```
let's make it explicit what is the default!

Comment (vasqu) [reply src/transformers/models/bart/modeling_bart.py]: You're too fast :D will change in a sec, wouldnt work either way with `attention_interface: Callable = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]` since eager is not registered in the interface

Comment (ArthurZucker) [reply src/transformers/models/bart/modeling_bart.py]: haha yep exactly

File: tests/test_modeling_common.py [hunk 12]
```diff
@@ -4513,13 +4513,28 @@ def test_flex_attention_with_grads(self):
             # TODO: raushan, fix for composite models after making VLMs support new attn API
             if not model_class._supports_flex_attn or self._is_composite:
                 self.skipTest(reason="This model does not support flex attention")
+
             config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
             config._attn_implementation = "flex_attention"
-            model = model_class(config).to(device=torch_device, dtype=torch.float16)
+            # Flex Attention can not use dropout
+            if hasattr(config, "attention_droput"):
```
Comment (vasqu) [inline tests/test_modeling_common.py:4520]: Need to fix this, typo (discovered during whisper

Comment (vasqu) [reply tests/test_modeling_common.py]: Rerunning flex tests to see which models will fail.

Comment (vasqu) [reply tests/test_modeling_common.py]: Seems like a lot of models fail on flex attention... Not sure if I should submit a torch issue. Disabled flex for now - I don't think it's a high priority atm.

File: src/transformers/integrations/flash_attention.py [hunk 13]
```diff
@@ -18,8 +21,29 @@ def flash_attention_forward(
...
+    head_mask: Optional[torch.Tensor] = None,
+    eager_fallback: Optional[Callable] = None,
     **kwargs,
 ) -> Tuple[torch.Tensor, None]:
+    if output_attentions or head_mask is not None:
+        logger.warning_once(
+            "Falling back to eager attention because `flash_attention_2` does not support"
+            " `output_attentions=True` or `head_mask`."
+        )
+        return eager_fallback(
```
Comment (ArthurZucker) [inline src/transformers/integrations/flash_attention.py:34]: if eager fallback is none we are failing hard! Let's maybe just emit the warning no? And return None output attentions

File: src/transformers/integrations/flex_attention.py [hunk 14]
```diff
@@ -213,23 +230,44 @@ def flex_attention_forward(
...
+    dropout: float = 0.0,
+    eager_fallback: Optional[Callable] = None,
     **kwargs,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
+    if output_attentions or head_mask is not None or dropout > 0:
+        logger.warning_once(
+            "Falling back to eager attention because `flex_attention` does not support"
+            " `output_attentions=True`, `head_mask`, or `dropout`."
+        )
+        return eager_fallback(
```
Comment (ArthurZucker) [inline src/transformers/integrations/flex_attention.py:243]: same here! And head mask is something we kinda deprecated so let's just return None prob

**Review by ArthurZucker — APPROVED** [review]
Kudos to you! Let's break a bit and warn without fallback WDYT? not adding a new arg!

**Comment by vasqu** [comment]
Another day, another merge conflict
